### PR TITLE
Add user and password authentication.

### PR DESCRIPTION
**Pull request by ahherrera**
Issue: https://github.com/owncloud/search_elastic/issues/246
Add user and password authentication.

### CODE REVIEW

**Comment by IljaN** [comment]
Please ensure that CI is green :)

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISE-latest-chrome-mysql8.0-php7.4-es7.10</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/search_elastic/1929/16/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiLimitSearches-latest-mysql8.0-php7.4-es7.10</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/search_elastic/1929/19/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiLimitSearches-master-mysql8.0-php7.4-es7.10</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/search_elastic/1929/18/1

**Comment by ahherrera** [comment]
@IljaN I have had problems deploying the PR. Can you tell me how to solve these problems?
Screen Shot 2022-02-11 at 8 55 50 AM

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISE-master-firefox-mysql8.0-php7.4-es7.10</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/search_elastic/1930/15/1

**Comment by IljaN** [comment]
Hmhh something cancels the pipeline... maybe a failing acceptance test?

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISE-latest-chrome-mysql8.0-php7.4-es7.10</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/search_elastic/1931/16/1

File: lib/SearchElasticConfigService.php [hunk 1]
```diff
@@ -215,4 +213,40 @@ public function parseServers($servers) {
...
 	}
+
+	/**
+	 * Parse server connection with user and password.
+	 *
+	 * @param string $serverPart
+	 * @param array $server
+	 */
+	private function parseServerWithUserPassAuthentication($serverPart, &$server) {
+		$sets = explode('@', $serverPart);
```
Comment (IljaN) [inline lib/SearchElasticConfigService.php:224]:
```suggestion
		$sets = \explode('@', $serverPart);
```

File: lib/SearchElasticConfigService.php [hunk 2]
```diff
@@ -215,4 +213,40 @@ public function parseServers($servers) {
...
+		$authenticationParameters = explode(':', $sets[0]);
+		$host = explode(':', $sets[1]);
```
Comment (IljaN) [inline lib/SearchElasticConfigService.php:226]:
```suggestion
		$host = \explode(':', $sets[1]);
```

File: lib/SearchElasticConfigService.php [hunk 3]
```diff
@@ -197,16 +197,14 @@ public function parseServers($servers) {
...
 			$server = [
 				'host' => 'localhost',
 				'port' => 9200
 			];
-			if (!empty($hostAndPort[0])) {
-				$server['host'] = $hostAndPort[0];
-			}
-			if (!empty($hostAndPort[1])) {
-				$server['port'] = (int)$hostAndPort[1];
+			if (str_contains($serverPart, '@')) {
```
Comment (IljaN) [inline lib/SearchElasticConfigService.php:204]: Is it possible to eliminate this if by merging the methods ``parseServerWithUserPassAuthentication`` and ``parseServerWithoutAuthentication`` in to one?

Comment (IljaN) [inline lib/SearchElasticConfigService.php:204]:
```suggestion
			if (\str_contains($serverPart, '@')) {
```

**Review by IljaN — CHANGES_REQUESTED** [review]
:+1: Just a few nitpicks. Don't merge the two parse functions if it's to complicated.

File: lib/SearchElasticConfigService.php (hunk 3 above)
Comment (ahherrera) [reply lib/SearchElasticConfigService.php:204]: The functionality is separated into two methods in order to facilitate new types of authentication in the future without damaging existing methods. Please let me know if it is necessary to merge them.

Comment (IljaN) [reply lib/SearchElasticConfigService.php]: :+1:

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiLimitSearches-latest-mysql8.0-php7.4-es7.10</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/search_elastic/1944/19/1

**Comment by mmattel** [comment]
I guess this is docs relevant, in case, can you pls file a docs issue.

@EParzefall fyi

**Pull request merged**